### ISSUE TITLE
Failure on landing+retiring workers 

### ISSUE THREAD

**Issue by baptiste-bonnaudet** [concourse/concourse#2195]
Hi there!

I'm trying to safely retire workers on our setup. 
When using the `concourse land-worker` the worker stays a few seconds in "landing" and then stays in "landed" but without waiting to drain and still shows containers on the worker with `fly workers`.

After that doing a `concourse retire-worker` exits 1 with a 500 error:
```
concourse retire-worker
{"timestamp":"1525286998.252250433","source":"retire-worker","message":"retire-worker.beacon.setting-up-beacon-runner","log_level":0,"data":{"session":"1"}}
{"timestamp":"1525286998.252328873","source":"retire-worker","message":"retire-worker.beacon.retire-worker","log_level":0,"data":{"session":"1"}}
{"timestamp":"1525286998.252354860","source":"retire-worker","message":"retire-worker.beacon.command-to-run","log_level":0,"data":{"cmd":"retire-worker","session":"1"}}
{"timestamp":"1525286998.308972836","source":"retire-worker","message":"retire-worker.beacon.beacon-client.keepalive.ok","log_level":0,"data":{"session":"1.1.1"}}
{"timestamp":"1525286998.309337854","source":"retire-worker","message":"retire-worker.beacon.ssh-forward-config","log_level":0,"data":{"bcForwardAddrRemote":"","gardenForwardAddrRemote":"","reaperForwardAddrRemote":"","session":"1"}}
{"timestamp":"1525286998.309979916","source":"tsa","message":"tsa.connection.channel.retire-worker.start","log_level":1,"data":{"remote":"10.28.4.1:59892","session":"136.1.1"}}
{"timestamp":"1525286998.319484234","source":"tsa","message":"tsa.connection.channel.retire-worker.bad-response","log_level":2,"data":{"remote":"10.28.4.1:59892","session":"136.1.1","status-code":500}}
{"timestamp":"1525286998.319707155","source":"tsa","message":"tsa.connection.channel.retire-worker.end","log_level":1,"data":{"remote":"10.28.4.1:59892","session":"136.1.1"}}
{"timestamp":"1525286998.319833994","source":"tsa","message":"tsa.connection.channel.retire-worker.failed-to-retire-worker","log_level":2,"data":{"error":"bad-response (500): HTTP/1.1 500 Internal Server Error\r\nContent-Length: 0\r\nDate: Wed, 02 May 2018 18:49:58 GMT\r\nX-Concourse-Version: 3.11.0\r\nX-Content-Type-Options: nosniff\r\nX-Download-Options: noopen\r\nX-Xss-Protection: 1; mode=block\r\n\r\n","remote":"10.28.4.1:59892","session":"136.1.1"}}
{"timestamp":"1525286998.320283413","source":"retire-worker","message":"retire-worker.beacon.failed-to-keep-session-alive","log_level":2,"data":{"error":"Process exited with status 1","session":"1"}}
Process exited with status 1
```

In ATC we have the logs:
```
{"timestamp":"1525286732.999515057","source":"atc","message":"atc.syncer.docker-cis_elasticsearch:radar.scan-resource.interval-runner.tick.find-worker.worker-for-container.garden-connection.retryable-http-client.retrying","log_level":1,"data":{"error":"worker 'concourse-workers-v5-t75b' is unreachable (state is 'landed')","failed-attempts":14,"pipeline-scoped-name":"docker-cis_elasticsearch:resource:docker-image-cis_elasticsearch","ran-for":"2m39.067441238s","session":"46.51.4.1.401.3.1.1.1"}}
```

Is it a bug or  am I using it wrong?

* Concourse version: 3.11
* Deployment type (BOSH/Docker/binary): ATC on Docker, Workers on VMs.
* Infrastructure/IaaS: GCP

**Comment by shashwathi** [comment]
@baptiste-bonnaudet : I think there should be more ATC logs from the worker retire API. Can you search in ATC logs for `failed-to-retire-worker`, `failed-finding-worker-to-retire` ?

**Comment by MiniJerome** [comment]
@shashwathi We upgraded to Concourse 3.13.0 and here what we have in ATC logs:

```
{"timestamp":"1525697586.645510674","source":"tsa","message":"tsa.connection.channel.retire-worker.start","log_level":1,"data":{"remote":"10.0.4.1:34464","session":"349.1.1"}}
{"timestamp":"1525697586.704839706","source":"atc","message":"atc.retiring-worker.failed-to-retire-worker","log_level":2,"data":{"error":"pq: new row for relation \"workers\" violates check constraint \"addr_when_running\"","session":"554923"}}
{"timestamp":"1525697586.705141306","source":"tsa","message":"tsa.connection.channel.retire-worker.bad-response","log_level":2,"data":{"remote":"10.0.4.1:34464","session":"349.1.1","status-code":500}}
{"timestamp":"1525697586.705368042","source":"tsa","message":"tsa.connection.channel.retire-worker.end","log_level":1,"data":{"remote":"10.0.4.1:34464","session":"349.1.1"}}
{"timestamp":"1525697586.705438137","source":"tsa","message":"tsa.connection.channel.retire-worker.failed-to-retire-worker","log_level":2,"data":{"error":"bad-response (500): HTTP/1.1 500 Internal Server Error\r\nContent-Length: 0\r\nDate: Mon, 07 May 2018 12:53:06 GMT\r\nX-Concourse-Version: 3.13.0\r\nX-Content-Type-Options: nosniff\r\nX-Download-Options: noopen\r\nX-Xss-Protection: 1; mode=block\r\n\r\n","remote":"10.0.4.1:34464","session":"349.1.1"}}
```

**Comment by MiniJerome** [comment]
@xtremerui yes, the issue is that a worker can't be retired when it is in `landed` state.

First, we land the worker, then we retire it but it doesn't works.

The worker stays `landing` few seconds and then stays in `landed` state until we prune it.

We must `prune-worker` the "landed" worker.

Here what we do and what we can see in logs:

```
Every 1.0s: fly -t stg ws

concourse-workers-v6-8zjt  3  linux  none  none  running  2.0
concourse-workers-v6-rwnn  2  linux  none  none  running  2.0
concourse-workers-v6-w717  4  linux  none  none  running  2.0
concourse-workers-v6-x4v3  4  linux  none  none  landing  2.0
concourse-workers-v7-tdq7  4  linux  none  none  running  2.0
```

Few seconds after that:

```
Every 1.0s: fly -t stg ws

concourse-workers-v6-8zjt  3  linux  none  none  running  2.0
concourse-workers-v6-rwnn  2  linux  none  none  running  2.0
concourse-workers-v6-w717  4  linux  none  none  running  2.0
concourse-workers-v6-x4v3  4  linux  none  none  landed   2.0
concourse-workers-v7-tdq7  4  linux  none  none  running  2.0

```

So we see some issues in ATC logs:

```
{"timestamp":"1525717044.993446350","source":"tsa","message":"tsa.connection.channel.channel-request","log_level":1,"data":{"remote":"10.142.0.4:50192","session":"404.1","type":"exec"}}
{"timestamp":"1525717045.006907940","source":"tsa","message":"tsa.connection.channel.land-worker.start","log_level":1,"data":{"remote":"10.142.0.4:50192","session":"404.1.1"}}
{"timestamp":"1525717045.006998777","source":"tsa","message":"tsa.connection.tcpip-forward.forwarding-tcpip","log_level":1,"data":{"remote":"10.142.0.4:50192","requested-bind-addr":"0.0.0.0:7777","session":"404.2"}}
{"timestamp":"1525717045.019978523","source":"tsa","message":"tsa.connection.tcpip-forward.forwarding-tcpip","log_level":1,"data":{"remote":"10.142.0.4:50192","requested-bind-addr":"0.0.0.0:7788","session":"404.3"}}
{"timestamp":"1525717045.033036470","source":"tsa","message":"tsa.connection.tcpip-forward.forwarding-tcpip","log_level":1,"data":{"remote":"10.142.0.4:50192","requested-bind-addr":"0.0.0.0:7799","session":"404.4"}}
{"timestamp":"1525717045.068044662","source":"tsa","message":"tsa.connection.channel.land-worker.end","log_level":1,"data":{"remote":"10.142.0.4:50192","session":"404.1.1"}}
{"timestamp":"1525717045.081331015","source":"tsa","message":"tsa.connection.tcpip-forward.failed-to-accept","log_level":2,"data":{"error":"accept tcp [::]:35858: use of closed network connection","remote":"10.142.0.4:50192","session":"404.4"}}
{"timestamp":"1525717045.081421375","source":"tsa","message":"tsa.connection.tcpip-forward.failed-to-accept","log_level":2,"data":{"error":"accept tcp [::]:32947: use of closed network connection","remote":"10.142.0.4:50192","session":"404.3"}}
{"timestamp":"1525717045.081459284","source":"tsa","message":"tsa.connection.tcpip-forward.failed-to-accept","log_level":2,"data":{"error":"accept tcp [::]:32847: use of closed network connection","remote":"10.142.0.4:50192","session":"404.2"}}
...
{"timestamp":"1525717108.966801167","source":"atc","message":"atc.syncer.go-adt-gcp-stg:radar.scan-resource.interval-runner.tick.find-worker.worker-for-container.garden-connection.retryable-http-client.retrying","log_level":1,"data":{"error":"worker 'concourse-workers-v6-x4v3' is unreachable (state is 'landed')","failed-attempts":7,"pipeline-scoped-name":"go-adt-gcp-stg:resource:code-base","ran-for":"47.195693739s","session":"36.1.13.1.126.3.1.1.1"}}
```

And if we try to trigger a build it stays in `pending` state until we `prune-worker` the `landed` worker - even though we have multiple `running` workers. Git resources are stucks too.

![screen shot 2018-05-07 at 2 23 46 pm](https://user-images.githubusercontent.com/26054609/39718742-3631ec92-5205-11e8-8842-f42696b2a27f.png)
![screen shot 2018-05-07 at 2 23 15 pm](https://user-images.githubusercontent.com/26054609/39718756-40c93ce6-5205-11e8-81f6-018fff11ae89.png)

And, when we try to `retire-worker`:

Workers logs:
```
{"timestamp":"1525717468.168936014","source":"retire-worker","message":"retire-worker.beacon.setting-up-beacon-runner","log_level":0,"data":{"session":"1"}}
{"timestamp":"1525717468.169015884","source":"retire-worker","message":"retire-worker.beacon.retire-worker","log_level":0,"data":{"session":"1"}}
{"timestamp":"1525717468.169051647","source":"retire-worker","message":"retire-worker.beacon.command-to-run","log_level":0,"data":{"cmd":"retire-worker","session":"1"}}
{"timestamp":"1525717468.324230909","source":"retire-worker","message":"retire-worker.beacon.beacon-client.keepalive.ok","log_level":0,"data":{"session":"1.1.1"}}
{"timestamp":"1525717468.336580515","source":"retire-worker","message":"retire-worker.beacon.ssh-forward-config","log_level":0,"data":{"bcForwardAddrRemote":"","gardenForwardAddrRemote":"","reaperForwardAddrRemote":"","session":"1"}}
{"timestamp":"1525717468.343023777","source":"tsa","message":"tsa.connection.channel.retire-worker.start","log_level":1,"data":{"remote":"10.142.0.4:50270","session":"407.1.1"}}
{"timestamp":"1525717468.402782679","source":"tsa","message":"tsa.connection.channel.retire-worker.bad-response","log_level":2,"data":{"remote":"10.142.0.4:50270","session":"407.1.1","status-code":500}}
{"timestamp":"1525717468.403362036","source":"tsa","message":"tsa.connection.channel.retire-worker.end","log_level":1,"data":{"remote":"10.142.0.4:50270","session":"407.1.1"}}
{"timestamp":"1525717468.403716803","source":"tsa","message":"tsa.connection.channel.retire-worker.failed-to-retire-worker","log_level":2,"data":{"error":"bad-response (500): HTTP/1.1 500 Internal Server Error\r\nContent-Length: 0\r\nDate: Mon, 07 May 2018 18:24:28 GMT\r\nX-Concourse-Version: 3.13.0\r\nX-Content-Type-Options: nosniff\r\nX-Download-Options: noopen\r\nX-Xss-Protection: 1; mode=block\r\n\r\n","remote":"10.142.0.4:50270","session":"407.1.1"}}
{"timestamp":"1525717468.410596848","source":"retire-worker","message":"retire-worker.beacon.failed-to-keep-session-alive","log_level":2,"data":{"error":"Process exited with status 1","session":"1"}}
Process exited with status 1
```

Logs from ATC:

```
{"timestamp":"1525717468.343023777","source":"tsa","message":"tsa.connection.channel.retire-worker.start","log_level":1,"data":{"remote":"10.142.0.4:50270","session":"407.1.1"}}
{"timestamp":"1525717468.343264103","source":"tsa","message":"tsa.connection.tcpip-forward.forwarding-tcpip","log_level":1,"data":{"remote":"10.142.0.4:50270","requested-bind-addr":"0.0.0.0:7777","session":"407.2"}}
{"timestamp":"1525717468.356048346","source":"tsa","message":"tsa.connection.tcpip-forward.forwarding-tcpip","log_level":1,"data":{"remote":"10.142.0.4:50270","requested-bind-addr":"0.0.0.0:7788","session":"407.3"}}
{"timestamp":"1525717468.369159222","source":"tsa","message":"tsa.connection.tcpip-forward.forwarding-tcpip","log_level":1,"data":{"remote":"10.142.0.4:50270","requested-bind-addr":"0.0.0.0:7799","session":"407.4"}}
{"timestamp":"1525717468.402350187","source":"atc","message":"atc.retiring-worker.failed-to-retire-worker","log_level":2,"data":{"error":"pq: new row for relation \"workers\" violates check constraint \"addr_when_running\"","session":"773886"}}
{"timestamp":"1525717468.402782679","source":"tsa","message":"tsa.connection.channel.retire-worker.bad-response","log_level":2,"data":{"remote":"10.142.0.4:50270","session":"407.1.1","status-code":500}}
{"timestamp":"1525717468.403362036","source":"tsa","message":"tsa.connection.channel.retire-worker.end","log_level":1,"data":{"remote":"10.142.0.4:50270","session":"407.1.1"}}
{"timestamp":"1525717468.403716803","source":"tsa","message":"tsa.connection.channel.retire-worker.failed-to-retire-worker","log_level":2,"data":{"error":"bad-response (500): HTTP/1.1 500 Internal Server Error\r\nContent-Length: 0\r\nDate: Mon, 07 May 2018 18:24:28 GMT\r\nX-Concourse-Version: 3.13.0\r\nX-Content-Type-Options: nosniff\r\nX-Download-Options: noopen\r\nX-Xss-Protection: 1; mode=block\r\n\r\n","remote":"10.142.0.4:50270","session":"407.1.1"}}
{"timestamp":"1525717468.416871786","source":"tsa","message":"tsa.connection.tcpip-forward.failed-to-accept","log_level":2,"data":{"error":"accept tcp [::]:43755: use of closed network connection","remote":"10.142.0.4:50270","session":"407.4"}}
{"timestamp":"1525717468.417208910","source":"tsa","message":"tsa.connection.tcpip-forward.failed-to-accept","log_level":2,"data":{"error":"accept tcp [::]:42711: use of closed network connection","remote":"10.142.0.4:50270","session":"407.3"}}
{"timestamp":"1525717468.417391300","source":"tsa","message":"tsa.connection.tcpip-forward.failed-to-accept","log_level":2,"data":{"error":"accept tcp [::]:37582: use of closed network connection","remote":"10.142.0.4:50270","session":"407.2"}}
...
{"timestamp":"1525717477.689275980","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.scan.find-worker.worker-for-container.garden-connection.retryable-http-client.retrying","log_level":1,"data":{"build-id":1789,"build-name":"23","error":"worker 'concourse-workers-v6-x4v3' is unreachable (state is 'landed')","failed-attempts":11,"input":"code-base","job_name":"unit-test_frontend","ran-for":"1m51.30465396s","resource":"code-base","session":"768925.1.1.2.221.1.1.1"}}
```


As we see, our builds are stucks and the worker isn't retired.
Fly still shows `landed` state for this worker.

```
concourse-workers-v6-8zjt  3  linux  none  none  running  2.0
concourse-workers-v6-rwnn  2  linux  none  none  running  2.0
concourse-workers-v6-w717  4  linux  none  none  running  2.0
concourse-workers-v6-x4v3  4  linux  none  none  landed   2.0
concourse-workers-v7-tdq7  4  linux  none  none  running  2.0
```

Now, our only solution to unblock our pipelines is to run `prune-worker`, and then everything goes fine and builds are triggered.

```
pruned 'concourse-workers-v6-x4v3'
```

```
{"timestamp":"1525717589.884659767","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.scan.find-worker.worker-for-container.garden-connection.retryable-http-client.retrying","log_level":1,"data":{"build-id":1789,"build-name":"23","error":"worker 'concourse-workers-v6-x4v3' is unreachable (state is 'landed')","failed-attempts":18,"input":"code-base","job_name":"unit-test_frontend","ran-for":"3m43.500035843s","resource":"code-base","session":"768925.1.1.2.221.1.1.1"}}
{"timestamp":"1525717605.911461830","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.scan.failed-to-lookup-created-container-in-garden","log_level":2,"data":{"build-id":1789,"build-name":"23","container":"b74438c6-5e0b-493a-7066-5cc803618b12","error":"Get http://127.0.0.1:8080/containers: worker concourse-workers-v6-x4v3 disappeared while trying to reach it","input":"code-base","job_name":"unit-test_frontend","resource":"code-base","session":"768925.1.1.2"}}
{"timestamp":"1525717605.911539555","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.scan.failed-to-initialize-new-container","log_level":2,"data":{"build-id":1789,"build-name":"23","error":"Get http://127.0.0.1:8080/containers: worker concourse-workers-v6-x4v3 disappeared while trying to reach it","input":"code-base","job_name":"unit-test_frontend","resource":"code-base","session":"768925.1.1.2"}}
{"timestamp":"1525717605.964525938","source":"atc","message":"atc.create-job-build.trigger-immediately.failed-to-start-next-pending-build-for-job","log_level":2,"data":{"error":"Get http://127.0.0.1:8080/containers: worker concourse-workers-v6-x4v3 disappeared while trying to reach it","job-name":"unit-test_frontend","job_name":"unit-test_frontend","session":"768925.1"}}
{"timestamp":"1525717615.069284439","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.starting","log_level":1,"data":{"build-id":1788,"build-name":"35","job_name":"unit-test","session":"768818.1.1"}}
{"timestamp":"1525717615.164622545","source":"atc","message":"atc.create-job-build.trigger-immediately.try-start-next-pending-build.running","log_level":1,"data":{"build":1788,"build-id":1788,"build-name":"35","job":"unit-test","job_name":"unit-test","pipeline":"go-adt-gcp-stg","session":"768818.1.1"}}
{"timestamp":"1525717617.938904524","source":"atc","message":"atc.syncer.go-adt-gcp-stg:scheduler.tick.scheduling.try-start-next-pending-build.starting","log_level":1,"data":{"build-id":1788,"build-name":"35","session":"36.2.23298.3.6"}}
```

**Comment by pn-santos** [comment]
> _a worker can be retired only when its state is landed or stalled. So you will need to land the worker first and then try retiring._

@xtremerui I though that when you retire a worker, concourse will first drain it (effectively landing it) and then retire it. According to the [docs](https://www.concourse-ci.org/worker-internals.html#RETIRING-table) that seems to be the behavior:

> _The ATC will wait for builds on the worker for jobs which aren't interruptible to finish, and remove the worker._

**Comment by baptiste-bonnaudet** [comment]
Pivotal is using BOSH to deploy Concourse, is there some logic in there for upgrading that we are missing in our non-BOSH setup?

**Comment by shashwathi** [comment]
@pn-santos : 

ATC keeps track of worker state; if the worker is `retired` before transitioning the state to `landed` or `stalled` DB constraint prevents it.

**Comment by baptiste-bonnaudet** [comment]
@shashwathi thanks for the answer! 

So if I understand correctly there's a binary to control workers thats not part of the concourse cli.
Does the `concourse land-worker` does the same `drain` than the worker binary? 

In our case it seems that the worker does not drain before getting in the `landed` state. Then other triggered builds stay pending and do not schedule on other running workers until the point where we prune the landed workers as the retire command is not working.

**Comment by shashwathi** [comment]
> So if I understand correctly there's a binary to control workers thats not part of the concourse cli.

Yes thats correct. The functionality however should be similar in terms of landing or retiring workers. `Drain` feature on worker binary was specifically developed for bosh purpose([more info about drain](https://bosh.io/docs/drain/)). 

> Does the concourse land-worker does the same drain than the worker binary

`land-worker` calls ATC api to land a worker. `drain` does little more than that, it is meant to cover multiple cases like 
1. If worker process is not present, `drain` will exit 0.
1. If ATC is not reachable `drain` will stop trying to land worker
1. If drain timeout has crossed, `drain` will delete the worker and many more

Outside of bosh scenario `drain` is not recommended, concourse/worker binary `land-worker` / `retire-worker` command should be sufficient.

The real problem is why builds are in pending state and not scheduling work on other workers. 

1. Are builds looking for tagged workers?

Ideally ATC should be looking for any `running` workers to allocate work. 

** Note ** 
You can either land or retire a running worker but you cannot retire a landed worker. 

Worker lifecycle: 
running --> landing --> landed -->(heartbeat again) --> running
running --> retiring --> retired(deleted from DB altogether)
running --> (no heartbeat) --> stalled --> (heartbeat) --> running ....

**Comment by prein** [comment]
Hi, sorry for resurrecting old issue and feel free to tell me if it's not related.
I'm running 4.2.1 on Kubernetes, installed using official Helm chart. After about a day of running, some workers (not all) go into stalled, stop responding to liveness probes resulting in kubernetes looping over restarting the worker pods.

Note that I'm not executing any worker related commands until I run prune-worker which solves the problem (temporarily).

The errors I've found in "web" container logs:
`{"timestamp":"1541759222.458709240","source":"tsa","message":"tsa.connection.keepalive","log_level":1,"data":{"remote":"10.38.0.3:55846","session":"1259","type":"keepalive"}}
{"timestamp":"1541759222.459047079","source":"tsa","message":"tsa.connection.channel.channel-request","log_level":1,"data":{"remote":"10.38.0.3:55846","session":"1259.1","type":"exec"}}
{"timestamp":"1541759222.459466934","source":"tsa","message":"tsa.connection.tcpip-forward.forwarding-tcpip","log_level":1,"data":{"remote":"10.38.0.3:55846","requested-bind-addr":"0.0.0.0:7777","session":"1259.2"}}
{"timestamp":"1541759222.459551573","source":"tsa","message":"tsa.connection.channel.retire-worker.start","log_level":1,"data":{"remote":"10.38.0.3:55846","session":"1259.1.1"}}
{"timestamp":"1541759222.460235596","source":"tsa","message":"tsa.connection.tcpip-forward.forwarding-tcpip","log_level":1,"data":{"remote":"10.38.0.3:55846","requested-bind-addr":"0.0.0.0:7788","session":"1259.3"}}
{"timestamp":"1541759222.463730574","source":"atc","message":"atc.retiring-worker.failed-to-retire-worker","log_level":2,"data":{"error":"pq: new row for relation \"workers\" violates check constraint \"addr_when_running\"","session":"27515"}}
{"timestamp":"1541759222.463845491","source":"tsa","message":"tsa.connection.channel.retire-worker.bad-response","log_level":2,"data":{"remote":"10.38.0.3:55846","session":"1259.1.1","status-code":500}}
{"timestamp":"1541759222.463988066","source":"tsa","message":"tsa.connection.channel.retire-worker.end","log_level":1,"data":{"remote":"10.38.0.3:55846","session":"1259.1.1"}}
{"timestamp":"1541759222.464131117","source":"tsa","message":"tsa.connection.channel.retire-worker.failed-to-retire-worker","log_level":2,"data":{"error":"bad-response (500): HTTP/1.1 500 Internal Server Error\r\nContent-Length: 0\r\nDate: Fri, 09 Nov 2018 10:27:02 GMT\r\nX-Concourse-Version: 4.2.1\r\nX-Content-Type-Options: nosniff\r\nX-Download-Options: noopen\r\nX-Xss-Protection: 1; mode=block\r\n\r\n","remote":"10.38.0.3:55846","session":"1259.1.1"}}
{"timestamp":"1541759222.464676142","source":"tsa","message":"tsa.connection.tcpip-forward.failed-to-accept","log_level":2,"data":{"error":"accept tcp [::]:43225: use of closed network connection","remote":"10.38.0.3:55846","session":"1259.3"}}
{"timestamp":"1541759222.464724302","source":"tsa","message":"tsa.connection.tcpip-forward.failed-to-accept","log_level":2,"data":{"error":"accept tcp [::]:37398: use of closed network connection","remote":"10.38.0.3:55846","session":"1259.2"}}`

and events from kubernetes

`  Normal   Started                18m (x4 over 22m)  kubelet, ops-qa-blue-kube-minion-0  Started container
  Normal   Killing                18m (x3 over 21m)  kubelet, ops-qa-blue-kube-minion-0  Killing container with id docker://concourse-worker:Container failed liveness probe.. Container will be killed and recreated.
  Warning  Unhealthy              18m (x4 over 22m)  kubelet, ops-qa-blue-kube-minion-0  Liveness probe failed: Still trying to clean up before starting concourse. 'fly prune-worker -w concourse-worker-1' might need to be called to force cleanup.
  Warning  FailedPreStopHook      12m (x5 over 21m)  kubelet, ops-qa-blue-kube-minion-0  Exec lifecycle hook ([/bin/sh -c while ! concourse retire-worker --name=${HOSTNAME} | grep -q worker-not-found; do
  sleep 5
done]) for Container "concourse-worker" in Pod "concourse-worker-1_concourse(56a51c23-e407-11e8-b6c7-02618d079e8a)" failed - error: command '/bin/sh -c while ! concourse retire-worker --name=${HOSTNAME} | grep -q worker-not-found; do
  sleep 5
done' exited with 137: Process exited with status 1
Process exited with status 1
Process exited with status 1
Process exited with status 1
Process exited with status 1
Process exited with status 1
Process exited with status 1
Process exited with status 1
Process exited with status 1
Process exited with status 1
Process exited with status 1
Process exited with status 1
, message: "Process exited with status 1\nProcess exited with status 1\nProcess exited with status 1\nProcess exited with status 1\nProcess exited with status 1\nProcess exited with status 1\nProcess exited with status 1\nProcess exited with status 1\nProcess exited with status 1\nProcess exited with status 1\nProcess exited with status 1\nProcess exited with status 1\n"
  Warning  BackOff  2m (x38 over 14m)  kubelet, ops-qa-blue-kube-minion-0  Back-off restarting failed container`

Perhaps related https://github.com/concourse/concourse/issues/1419#issuecomment-427852304

**Comment by insider89** [comment]
Hello @MiniJerome. Did you find the workaround? I have the same problem. First, I am landing worker, then, when the state is `landed` try to retire it, but got the error. I want to set up system script which would delete worker from web node when worker if going to terminate.

**Comment by marco-m** [comment]
As explained above in https://github.com/concourse/concourse/issues/2195#issuecomment-387522586:

```
Note
You can either land or retire a running worker but you cannot retire a landed worker.

Worker lifecycle:
running --> landing --> landed -->(heartbeat again) --> running
running --> retiring --> retired(deleted from DB altogether)
running --> (no heartbeat) --> stalled --> (heartbeat) --> running ....
```

**Comment by zacharyw** [comment]
I'm experiencing the same issue described by @prein .

Noticed after a couple of days that none of our concourse scheduled tasks were running. After investigating we saw that the concourse workers had restarted thousands of times, and the pods were stuck in `CrashLoopBackOff`.

I redeployed the helm chart, which didn't resolve the issue, so I manually pruned the workers, and that fixed it.

My concern is that I had to go in and fix this manually, and I'm not sure what led to the problem so I don't know how to prevent it in the future.
